netty/src/main/java/io/grpc/transport/netty/Http2Negotiator.java

```diff
@@ -112,168 +97,252 @@ public static ChannelHandler serverTls(SSLEngine sslEngine) {
+            // If an error occurred during the handshake, throw it to the pipeline.
+            if (future.isSuccess()) {
+              completeFuture.get();
+            } else {
+              future.get();
+            }
+          }
+        });
+    return new BufferUntilTlsNegotiatedHandler(sslHandler, handler);
   }
 
   /**
-   * Create a plaintext upgrade negotiation for HTTP/1.1 to HTTP/2.
+   * Returns a {@link ChannelHandler} that ensures that the HTTP to HTTP/2 upgrade is performed,
+   * the {@code handler} is added to the pipeline writes to the {@link Channel} may happen
+   * immediately, even before the upgrade is complete.
    */
-  public static Negotiation plaintextUpgrade(Http2ConnectionHandler handler) {
+  public static ChannelHandler plaintextUpgrade(Http2ConnectionHandler handler) {
     // Register the plaintext upgrader
     Http2ClientUpgradeCodec upgradeCodec = new Http2ClientUpgradeCodec(handler);
     HttpClientCodec httpClientCodec = new HttpClientCodec();
     final HttpClientUpgradeHandler upgrader =
         new HttpClientUpgradeHandler(httpClientCodec, upgradeCodec, 1000);
-    final UpgradeCompletionHandler completionHandler = new UpgradeCompletionHandler();
-    final ChannelInitializer<Channel> initializer = new ChannelInitializer<Channel>() {
-      @Override
-      public void initChannel(Channel ch) throws Exception {
-        ch.pipeline().addLast(upgrader);
-        ch.pipeline().addLast(completionHandler);
+    return new BufferUntilHttpUpgradeCompleteHandler(upgrader);
+  }
+
+  /**
+   * Returns a {@link ChannelHandler} that ensures that the {@code handler} is added to the
+   * pipeline writes to the {@link Channel} may happen immediately, even before it is active.
+   */
+  public static ChannelHandler plaintext(final ChannelHandler handler) {
+    return new BufferUntilChannelActiveHandler(handler);
+  }
+
+  /**
+   * Buffers all writes until either {@link #flushBufferedAndRemove(ChannelHandlerContext)} or
+   * {@link #failBufferedAndClose(ChannelHandlerContext)} is called. This handler allows us to
+   * write to a {@link Channel} before we are allowed to write to it officially i.e.
+   * before it's active or the TLS Handshake is complete.
+   */
+  private abstract static class BufferWritesHandler extends ChannelDuplexHandler {
+
+    private ChannelHandler[] handlers;
+    private Queue<ChannelWrite> bufferedWrites = new ArrayDeque<ChannelWrite>();
+    private boolean flushed;
+
+    /**
+     * @param handlers the ChannelHandlers are added to the pipeline on channelRegistered and
+     *                 before this handler.
+     */
+    BufferWritesHandler(ChannelHandler... handlers) {
+      this.handlers = handlers;
+    }
+
+    @Override
+    public void channelRegistered(ChannelHandlerContext ctx) throws Exception {
+      /**
+       * This check is necessary as a channel may be registered with different eventloops during
```

Comment:
Ok that's starting to clear things up, thanks. So your comment:

> We can't set up the pipeline by specifying a ChannelInitializer in the client bootstrap, because the initializer is not executed in the calling thread but on channelRegistered in the eventloop thread

.. really means that the `ChannelInitializer.initChannel` is called in the eventloop?  It should still be added to the pipeline in the calling thread, right? Or is there some awful Netty hack going on?  Either way, the reason you can't extend `ChannelInitializer` is that the initializer's `channelRegistered` is final (so you can't override) and it immediately removes itself from the pipeline ... not what you want here.